### ISSUE TITLE
Update caching mechanisms to support refresh

### ISSUE THREAD

**Issue by david-rhodes** [mapbox/mapbox-unity-sdk#261]
Two thoughts:
1. Pass in a `ForceRefresh` bool to use web request, regardless if tile is cached
2. Fetch tile timestamp and compare to timestamp of cached tile--use web request if tile is out of date

Some background: https://github.com/mapbox/mapbox-unity-sdk/issues/251

**Comment by david-rhodes** [comment]
Another reference: https://github.com/mapbox/mapbox-unity-sdk/issues/212

**Comment by david-rhodes** [comment]
Another aspect to consider: specific types of tiles should never be cached (`traffic`, for example). Should we pass in a boolean when making tile requests that represents if it should be a candidate for caching, or not?

**Comment by BergWerkGIS** [comment]
Got a first experimental version working (branch [`auto-cache-refresh`](https://github.com/mapbox/mapbox-unity-sdk/compare/auto-cache-refresh?diff=split&name=auto-cache-refresh)) and would like to get your input on implementation and design before I continue working on it.
It's still `[wip]` so e.g. force inserting is not yet implemented.

----

> 1. Pass in a ForceRefresh bool to use web request, regardless if tile is cached

I opted for auto refreshing cache (no `ForceRefresh bool +`) that checks if a newer version is available - via `HEADER` request instead of a full `GET` request.

> 2. Fetch tile timestamp and compare to timestamp of cached tile

Unfortunately that does not work as not all our tile APIs populate the [`last-modified` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified).
I've settled on using [`etag` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag).

----

## current workflow(s) 

**tile not in cache:**
* download
* put into all actives caches
* return via callback

**tile in one of the caches:**
* immediately return cached tile via callback to get fast user experience
* _afterwards_ issue HEAD request and compare `etag` with cached `etag`
  * same `etag`: propagate to all other caches, but _**don't** force_ insert if exists
  * different `etag`:
    * download new tile
    * **force** insert into all caches

## issues with the current workflow(s)

* every tile, even when loaded from cache causes a `HEAD` request to be sent!
  * **we should verify that this doesn't count as a full tile request/mapview** @isiyu 
* I went for smooth user experience by always immediately returning a tile if it exists in a cache. The check if it is outdated happens afterwards.
  During a session a user might get to see an outdated tile first and only after he pans away and revisits the same location the new tile, which has been fetched in the background in the meantime, will be visible
* Because of previous point and our current implementation it is not possible to cancel a tile update request after the cached tile has been returned.

-----

Questions/suggestions/ideas?

**Comment by david-rhodes** [comment]
@BergWerkGIS Is there any sort of event or callback clients can subscribe to when a tile update request has been sent/received? An `ITileProvider` implementation could use this information to cleanly update the loaded (cached) tile (in view).

Have you considered ways of disabling caching for certain types of tiles/Map Ids?

**Comment by BergWerkGIS** [comment]
> Is there any sort of event or callback clients can subscribe to when a tile update request has been sent/received?

Currently not, as the callback passed in is used for returning the previously cached tile.
Maybe just call that again and see what happens.

> An ITileProvider implementation could use this information to cleanly update the loaded (cached) tile (in view).

Good thought, that would be the best way to do it.

**Comment by david-rhodes** [comment]
> Currently not, as the callback passed in is used for returning the previously cached tile.
Maybe just call that again and see what happens

Perhaps we should expose the caching object through a singleton or something? This way, we could insert new events and any client code could easily subscribe to them. Or, we could pass them up to `MapboxAcccess`?

**Comment by Irfanraza110** [comment]
Hi, currently i'm using unity 5.6 and mapbox sdk 1.0.1, so for force refresh feature should i upgrade it to 1.0.2 or should i copy and paste these script.

**Comment by david-rhodes** [comment]
@Irfanraza110 We recommend staying as current as possible as we can usually debug issues quicker this way. We will be releasing v1.3 soon. That said, the scripts associated with this ticket are not final and have not even been reviewed in detail, yet. I would not recommend porting them to your project unless you are comfortable with the issues listed above and willing to deal with conflicts when we do release a proper implementation in the future.

**Comment by Irfanraza110** [comment]
@david-rhodes  Any workaround would you suggest for me as I'm stuck in the middle of my project.
I only wanted to update my tile once my style is updated.I'm creating a mobile application for Android and iPhone, and I have also used these auto-refreshing script but it won't affect in my application. my map only updated when I clear cache from the editor.

**Comment by Alexphauge** [comment]
@david-rhodes will this feature be in 1.3? We're also looking forward to this, as our own implementation isn't working very well. :)

**Comment by david-rhodes** [comment]
Unfortunately this feature will not be in 1.3. 

If the idea is to use this for a published application before we get around to a proper implementation, I wonder if a temporary work around would be to use Unity Remote Settings to inform clients that they should clear their local caches?

**Comment by Irfanraza110** [comment]
For now the only solution would be this..we we're thinking about this solution from quite while..and i tried to clear cache with MapboxAccess.instance.ClearCache(); but it's not working on my Android application..?? Can you help me with some code...it would be really appreciated.
Thanks

**Comment by david-rhodes** [comment]
`ClearCache` will only clear those MapIds and StyleUrls that `MapboxAccess` knows about. In other words, if you have not requested a tile with those ids before (in your current application session), this will result in a no-op.

I suggest you ensure you've requested a sample for each Id/style you want to clear (maybe just one tile, for example), and then call `ClearCache`. Then, you will need to initialize your map. I know it's hacky--and I'm sorry. We'll try to improve for the future.

**Comment by Irfanraza110** [comment]
@david-rhodes I get notified when my map is updated but how can I know whether my cached tile is updated or not.if I clear cache when the map is updated then I unnecessarily need to clear cache even when my tile is not updated. so to check that my tile is updated or not i need to create a request to say to check my tile is updated or not as @BergWerkGIS says in his code.

`public IAsyncRequest Request(
			string uri
			, Action<Response> callback
			, int timeout = 10
			, CanonicalTileId tileId = new CanonicalTileId()
			, string mapId = null
		)
		{

			if (string.IsNullOrEmpty(mapId))
			{
				throw new Exception("Cannot cache without a map id");
			}

			CacheItem cachedItem = null;

			// go through existing caches and check if we already have the requested tile available
			foreach (var cache in _caches)
			{
				cachedItem = cache.Get(mapId, tileId);
				if (null != cachedItem)
				{
					break;
				}
			}

			var uriBuilder = new UriBuilder(uri);
			if (!string.IsNullOrEmpty(_accessToken))
			{
				string accessTokenQuery = "access_token=" + _accessToken;
				if (uriBuilder.Query != null && uriBuilder.Query.Length > 1)
				{
					uriBuilder.Query = uriBuilder.Query.Substring(1) + "&" + accessTokenQuery;
				}
				else
				{
					uriBuilder.Query = accessTokenQuery;
				}
			}
			string finalUrl = uriBuilder.ToString();


			// if tile was available call callback with it, propagate to all other caches and check if a newer one is available
			if (null != cachedItem)
			{
				callback(Response.FromCache(cachedItem.Data));

				IAsyncRequestFactory.CreateRequest(
					finalUrl,
					(Response headerOnly) =>
					{
						//just bail on error, we've returned the cached tile already
						if (headerOnly.HasError)
						{
							return;
						}

						//UnityEngine.Debug.LogFormat("{0} : {1}", cachedItem.ETag, headerOnly.Headers["ETag"]);
						// data from cache is the same as on the web, propagate to all other caches but don't force insert via cache.add()
						if (cachedItem.ETag.Equals(headerOnly.Headers["ETag"]))
						{
							foreach (var cache in _caches)
							{
								cache.Add(mapId, tileId, cachedItem, false);
							}
						}
						else
						{
							UnityEngine.Debug.LogWarningFormat("updating cached tile {0}", tileId);
							// request updated tile, don't pass callback as we've responsed with the previously cached tile
							requestTileAndCache(finalUrl, mapId, tileId, timeout, null);
						}
					}
					, timeout
					, HttpRequestType.Head
				);

				return new MemoryCacheAsyncRequest(uri);
			}
			else
			{
				// requested tile is not in any of the caches yet, get it
				return requestTileAndCache(finalUrl, mapId, tileId, timeout, callback);
			}
		}`

than if i get into that condition that say "updating cached tile" in else than i need to reload the map or reinitialize the map. by using ReloadMap.cs or by saying AbstractTileProvider method initialize. correct me if I'm wrong.

Thanks.

**Comment by david-rhodes** [comment]
@Irfanraza110 Can you explain to me your use-case, please? I don't understand why you would need to clear your cache so frequently that this would be a problem. If it's just for development purposes, you can clear cache by deleting the application.

**Comment by Irfanraza110** [comment]
@david-rhodes sorry for the late replay We are creating a game in which user has to collect something by going somewhere,
like we are doing in Pokemon go. till now we are doing that we put some point on map and user has to collect those point but by putting something on a map it was not refreshing like we discussed earlier.So we found another way of doing this like we will send lat/long of that point to particular client and client will show those point. I think we were misusing map box what you say is we are going on right way..? and sorry for hijacking this thread.
Thanks.
